src/main.cpp

```diff
@@ -862,16 +870,45 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state,
     }
 }
 
-bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidationState &state)
+bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidationState &state, uint32_t consensusBranchId)
 {
     // Basic checks that don't depend on any context
 
-    // Check transaction version
-    if (tx.nVersion < MIN_TX_VERSION) {
+    // Check min and max tx version of pre-Overwinter and Overwintered transactions.
+    if (tx.fOverwintered) {
+        if (tx.nVersion > CTransaction::OVERWINTER_MAX_CURRENT_VERSION || tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
+            return state.DoS(100, error("CheckTransaction(): unknown overwintered version"),
+                REJECT_INVALID, "unknown-overwintered-version");
+        }
+    }
+    else if (tx.nVersion < MIN_TX_VERSION) {
         return state.DoS(100, error("CheckTransaction(): version too low"),
                          REJECT_INVALID, "bad-txns-version-too-low");
     }
 
+    // When checking the transaction against the consensus rules for the Overwinter branch, any transaction
+    // with version format >= 3 must have the fOverwintered flag set.
+    // With Sprout consensus rules, where the branch id is 0, we do not need to perform this check as the
+    // Sprout epoch allows for transactions with version number >= 3 (without setting fOverwintered flag).
+    if (NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId == consensusBranchId) {
+        if (!tx.fOverwintered && tx.nVersion >= 3) {
+            return state.DoS(100, error("CheckTransaction(): version requires overwinter flag to be set"),
+                REJECT_INVALID, "overwinter-flag-not-set");
+        }
+    }
+
+    // Check transaction expiry height is within valid range
+    if (tx.fOverwintered && tx.nVersion == 3 && tx.nExpiryHeight > TX_EXPIRY_HEIGHT_THRESHOLD) {
+        return state.DoS(100, error("CheckTransaction(): expiry height is too high"),
+                         REJECT_INVALID, "overwinter-expiry-height-too-high");
+    }
+
+    // Check Overwintered tx version group id matches what Zcash has assigned to v3 tx format version.
+    if (tx.fOverwintered && tx.nVersion == 3 && tx.nVersionGroupId != V3_VERSION_GROUP_ID) {
```

Comment:
Check this before the `nExpiryHeight` check, because if `nVersionGroupId` is incorrect then we can't make any assumptions about the contents of `nExpiryHeight`.